File: c/medium/find_kth_bit_in_nth_binary_string.c
#include <stdio.h>

// Recursive function
char findKthBit(int n, int k) {
    // Base case: S1 = "0"
    if (k == 1)
        return '0';

    // Compute bit length of k
    int bit_size = 0;
    int temp = k;
    while (temp > 0) {
        bit_size++;
        temp >>= 1;
    }

    // Mirror index across power-of-two boundary
    int mirrored_index = (1 << bit_size) - k;

    // Symmetry center
    if (mirrored_index == k)
        return '1';

    // Recursive call
    char mirrored_bit = findKthBit(n, mirrored_index);

    // Invert bit
    return (mirrored_bit == '1') ? '0' : '1';
}

File: rust/medium/find_kth_bit_in_nth_binary_string.rs
fn find_kth_bit(n: i32, k: i32) -> char {
    // Base case
    if k == 1 {
        return '0';
    }

    // bit_length equivalent
    let bit_size = 32 - (k as u32).leading_zeros();

    // Mirror index
    let mirrored_index = (1 << bit_size) - k;

    // Symmetry center
    if mirrored_index == k {
        return '1';
    }

    // Recursive lookup
    let mirrored_bit = find_kth_bit(n, mirrored_index);

    // Invert
    if mirrored_bit == '1' { '0' } else { '1' }
}